api/v1alpha4/clusterclass_webhook.go

```diff
@@ -194,22 +200,95 @@ func (in *ClusterClass) validate(old *ClusterClass) error {
 		)
 	}
 
-	for _, class := range in.Spec.Workers.MachineDeployments {
-		for _, oldClass := range old.Spec.Workers.MachineDeployments {
-			if class.Class == oldClass.Class && !reflect.DeepEqual(class, oldClass) {
-				allErrs = append(allErrs,
-					field.Invalid(
-						field.NewPath("spec", "workers", "machineDeployments"),
-						class,
-						"cannot be changed.",
-					),
-				)
-			}
+	return allErrs
+}
+
+func (r LocalObjectTemplate) validate(namespace string, pathPrefix *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	// check if a name is provided
+	if r.Ref.Name == "" {
+		allErrs = append(allErrs,
+			field.Invalid(
+				pathPrefix.Child("ref", "name"),
+				r.Ref.Name,
+				"cannot be empty",
+			),
+		)
+	}
+
+	// validate if namespace matches the provided namespace
+	if namespace != "" && r.Ref.Namespace != namespace {
+		allErrs = append(
+			allErrs,
+			field.Invalid(
+				pathPrefix.Child("ref", "namespace"),
+				r.Ref.Namespace,
+				fmt.Sprintf("must be '%s'", namespace),
+			),
+		)
+	}
+
+	// check if kind is a template
+	if len(r.Ref.Kind) <= len(TemplateSuffix) || !strings.HasSuffix(r.Ref.Kind, TemplateSuffix) {
+		allErrs = append(allErrs,
+			field.Invalid(
+				pathPrefix.Child("ref", "kind"),
+				r.Ref.Kind,
+				fmt.Sprintf("kind must be of form '<name>%s'", TemplateSuffix),
+			),
+		)
+	}
+
+	// check if apiVersion is valid
+	if gv, err := schema.ParseGroupVersion(r.Ref.APIVersion); err != nil || gv.Empty() {
+		if err != nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					pathPrefix.Child("ref", "apiVersion"),
+					r.Ref.APIVersion,
+					fmt.Sprintf("must be a valid apiVerison: %v", err),
+				),
+			)
+		} else {
+			allErrs = append(allErrs,
+				field.Invalid(
+					pathPrefix.Child("ref", "apiVersion"),
+					r.Ref.APIVersion,
+					"cannot be empty",
+				),
+			)
 		}
 	}
 
-	if len(allErrs) > 0 {
-		return apierrors.NewInvalid(GroupVersion.WithKind("ClusterClass").GroupKind(), in.Name, allErrs)
+	return allErrs
+}
+
+// classes returns the set of MachineDeployment class names.
+func (w WorkersClass) classes() sets.String {
```

Comment:
```suggestion
func (w WorkersClass) classNames() sets.String {
```
up to you. Just resolve if you like the current version better